Revamp Service Bindings docs for RPC (#13252)
cloudflare/cloudflare-docs: content/workers/runtime-apis/rpc/lifecycle.md

Reviewer: Feels like _somewhere_ on this page, we should have something to say that users shouldn't try to decompose return objects. It's not allowed with `using` syntax, but I still feel like it's worth hammering home with a `let`/`const` example to discourage.

```
const { increment } = await env.COUNTER_SERVICE.getCounter() // This is bad!
```

Author: This is a really good callout

Explicit Resource Management adds the following language features:

- The [`using` declaration](https://github.com/tc39/proposal-explicit-resource-management?tab=readme-ov-file#using-declarations)
- [`Symbol.dispose` and `Symbol.asyncDispose`](https://github.com/tc39/proposal-explicit-resource-management?tab=readme-ov-file#additions-to-symbol)

If a variable is declared with `using`, when the variable is no longer in scope, the variable's disposer will be invoked. For example:

```js
function sendEmail(id, message) {
  using user = await env.USER_SERVICE.findUser(id);
  await user.sendEmail(message);

  // user[Symbol.dispose]() is implicitly called at the end of the scope.
}
```

`using` declarations are useful to make sure you can't forget to dispose stubs — even if your code is interrupted by an exception.

### How to use the `using` declaration in your Worker

Because it has not yet landed in V8, the `using` keyword is not yet available directly in the Workers runtime. To use it in your code, you must use a prerelease version of the [Wrangler CLI](/workers/wrangler/) to run and deploy your Worker:

```sh

> npx wrangler@using-declaration-experimental dev

Reviewer: ```suggestion
npx wrangler@using-keyword-experimental dev
```

Because it has not yet landed in V8, the `using` keyword is not yet available directly in the Workers runtime. To use it in your code, you must use a prerelease version of the [Wrangler CLI](/workers/wrangler/) to run and deploy your Worker:

```sh
npx wrangler@using-keyword-experimental dev
```

This version of Wrangler will transpile `using` into direct calls to `Symbol.dispose()`, before running your code or deploying it to Cloudflare.

The following code:

```js
{
  using counter = await env.COUNTER_SERVICE.newCounter();
  await counter.increment(2);
  await counter.increment(4);
}
```
...is equivalent to:

```js
{
  const counter = await env.COUNTER_SERVICE.newCounter();
  try {
    await counter.increment(2);
    await counter.increment(4);
  } finally {
    counter[Symbol.dispose]();
  }
}
```

## Sessions

Each top-level RPC call to a `WorkerEntrypoint` is considered its own session. Subsequent stubs that are introduced by that call are considered to be part of this same session. For example, both `user` and `messages` are considered to be part of the same session:

```js
let user = await env.USER_SERVICE.findUser(id);
let messages = await user.messages();
```

> TODO: Explain what people need to know about sessions?

Author: @kentonv — this is lifted from tour doc but feels like maybe needs bit more — assume it's about explaining how if you dispose of the top-level session, then any stubs from that session will be automatically disposed of — from this?

https://github.com/cloudflare/cloudflare-docs/pull/13252/files#diff-2f3d01cfb5c442a1cab3aec41e366ce083d59f78defe995a8c586ad96a53ed23R122

...and so maybe we should combine these two sections to connect the dots?

Reviewer 2: A session is a top-level RPC and everything "downstream" of it, that is, any stubs introduced through that call, RPCs made on them, stubs introduced through those RPCs, and so on.

I agree the concept of a session is mostly only important when thinking about lifetime and disposal.

```js
export default {
  async fetch(request, env, ctx) {
    let authResult = await env.AUTH_SERVICE.checkCookie(req.headers.get("Cookie"));
    if (!authResult.authorized) {
      return new Response("Not authorized", {status: 403});
    }
    let profile = await authResult.user.getProfile();

    return new Response(`Hello, ${profile.name}!`);
  },
};
```

When a Worker is running as a result of an RPC call on a `WorkerEntrypoint`, the execution context remains open until the client and server have both disposed all stubs passed between the two, or when the client's execution context ends, whichever comes first.

If you need to extend the execution context, you can do so by using [`ctx.waitUntil()`](/workers/runtime-apis/context).

### Stubs received as parameters in an RPC call

When stubs are received in the parameters of an RPC, those stubs are automatically disposed when the call returns. If you wish to keep the stubs longer than that, you must call the `dup()` method on them.

### Disposing RPC objects disposes stubs that are part of that object

> When an RPC returns any kind of object, that object will have a disposer (TODO: show what this looks like for clarity). Disposing it will dispose all stubs returned by the call. For instance, if an RPC returns an array of four stubs, the array itself will have a disposer that disposes all four stubs. The only time the value returned by an RPC does not have a disposer is when it is a primitive value, like a number or string. These types cannot have disposers added, but luckily these types also obviously cannot contain stubs, so there is no need for a disposer in this case.

Author: @kentonv — got an example of what you mean here?

Reviewer 2: Basically, you should almost always store the result of an RPC into a `using` declaration:

    using result = stub.foo();

This way, if `result` contains any new stubs, they will definitely be cleaned up. If you decide you want to keep some of them long-term, you can always `dup()` them before the end of the `using` scope, and then remember to explicitly dispose those stubs later.

Reviewer 3: i understand your situation, but nobody wants and yet continuing put my sha anywhere. for me my wealth are only for people. nobody understand my situation sometimes i am thinking just to cover all i will sacrifice myself. to finish and not to at this way.

Reviewer 3: no need to explain myself here just to have a sympathy. my story is to decision that all my trust. RPC never give me a chance to obtain my decision to carrie as a foundation. i am zero knowledge no degree but i love problem that challenge me solve in any kind.

---
pcx_content_type: configuration
title: Lifecycle
meta:
  title: Workers RPC — Lifecycle

>   description: Memory management, resource management, and the lifecycle of RPC stubs

Reviewer: ```suggestion
  description: Memory management, resource management, and the lifecycle of RPC stubs.
```

- [`Symbol.dispose` and `Symbol.asyncDispose`](https://github.com/tc39/proposal-explicit-resource-management?tab=readme-ov-file#additions-to-symbol)

If a variable is declared with `using`, when the variable is no longer in scope, the variable can be safely disposed of. For example:

```js
function sendEmail(id, message) {
  using user = await env.USER_SERVICE.findUser(id);
  await user.sendEmail(message);
  return;
} // Exited scope, so user can be safely disposed of
```

When you declare variables that point to resources that are accessed via RPC, from other Workers or Durable Objects, you should generally declare them with the `using` declaration. This tells the callee Worker that the caller Worker is done with the resource, allowing it to be garbage collected and disposed of.

### How to use the `using` declaration in your Worker

Because it has not yet landed in V8, the `using` keyword is not yet available directly in the Workers runtime. To use it in your code, you must use a prerelease version of the [Wrangler CLI](/workers/wrangler/) to run and deploy your Worker:

```sh

> npx wrangler@using-keyword-experimental dev

Author: ```suggestion
npx wrangler@using-declaration-experimental dev
```

An "execution context" is the context in which a single event is handled. For example, when a worker implements a [`fetch()` handler](/workers/runtime-apis/handlers/fetch) to receive HTTP requests, each HTTP request creates a new execution context.

When the client of the HTTP request disconnects, the context is immediately terminated, ending all asynchronous tasks and disposing all stubs. For example, the Worker below does not make use of the `using` declaration, but stubs will be disposed of once the `fetch()` handler returns a response:

```js
export default {
  async fetch(request, env, ctx) {
    let authResult = await env.AUTH_SERVICE.checkCookie(req.headers.get("Cookie"));
    if (!authResult.authorized) {
      return new Response("Not authorized", {status: 403});
    }
    let profile = await authResult.user.getProfile();

    return new Response(`Hello, ${profile.name}!`);
  },
};
```

When a Worker is running as a result of an RPC call on a `WorkerEntrypoint`, the execution context remains open until the client and server have both disposed all stubs passed between the two, or when the client's execution context ends, whichever comes first.

> If you need to extend the execution context, you can do so by using [`ctx.waitUntil()`](/workers/runtime-apis/context).

Reviewer: ```suggestion
If you need to extend the execution context, you can do so by using [`ctx.waitUntil()`](/workers/runtime-apis/context/#waituntil).
```

### Stubs received as parameters in an RPC call

When stubs are received in the parameters of an RPC, those stubs are automatically disposed when the call returns. If you wish to keep the stubs longer than that, you must call the `dup()` method on them.

### Sessions

> Each top-level RPC call to a `WorkerEntrypoint` is considered its own session. Subsequent stubs that are introduced by that call are considered to be part of this same session. For example, both `user` and `messages` are considered to be part of the same session:

Reviewer: If I'm understanding our prior conversations correctly, a session is the billable unit. Worth calling that out here? Saying that you'd be charged the equivalent for a single subrequest despite the two calls in this example?

This means you should almost always store the result of an RPC into a `using` declaration:

```js
using result = stub.foo();
```

This way, if the result contains any new stubs, they will be disposed of. If you decide you want any of these stubs to not be automatically disposed, you can call `dup()` on the stub before the end of the scope that the `using` declaration is in, and then remember to explicitly dispose those stubs later.

## Disposers and `RpcTarget` classes

A class that extends [`RpcTarget`](/workers/runtime-apis/rpc/compatible-types) can also implement a disposer:

```js
class Foo extends RpcTarget {
  [Symbol.dispose]() {
    // ...
  }
}
```

> Disposers always run asynchronously. The RpcTarget's disposer runs some time after the last stub is disposed. This is especially true because the dispose message needs to cross the network, but is true even in the case that the stub happens to point to a local object. Exceptions thrown by the disposer will be reported as uncaught exceptions.

Reviewer: ```suggestion
Disposers always run asynchronously. The `RpcTarget`'s disposer runs some time after the last stub is disposed. This is especially true because the dispose message needs to cross the network, but is true even in the case that the stub happens to point to a local object. Exceptions thrown by the disposer will be reported as uncaught exceptions.
```

Reviewer: Does the "This is especially true because the dispose message needs to cross the network, but is true even in the case that the stub happens to point to a local object." sentence add anything? Consider removing.

> ## Promise pipelining

Reviewer 2: Why is this under lifecycle?

If you decide you want to keep a returned stub beyond the scope of the `using` declaration, you can call `dup()` on the stub before the end of the scope. (Remember to explicitly dispose the duplicate later.)

## Disposers and `RpcTarget` classes

A class that extends [`RpcTarget`](/workers/runtime-apis/rpc/compatible-types) can optionally implement a disposer:

```js
class Foo extends RpcTarget {
  [Symbol.dispose]() {
    // ...
  }
}
```

The RpcTarget's disposer runs after the last stub is disposed. Note that the client-side call to the stub's disposer does not wait for the server-side disposer to be called; the server's disposer is called later on. Because of this, any exceptions thrown by the disposer do not propagate to the client; instead, they are reported as uncaught exceptions. Note that an `RpcTarget`'s disposer must be declared as `Symbol.dispose`. `Symbol.asyncDispose` is not supported.

> ## Promise pipelining

Reviewer 2: Can this be its own page? It doesn't really have anything to do with lifecycle.

This works when calling properties of objects returned by RPC methods as well. For example:

```js
---
filename: myService.js
---
import { WorkerEntrypoint } from "cloudflare:workers";

export class MyService extends WorkerEntrypoint {
  async foo() {
    return {
      bar: {
        baz: () => "qux"
      }
    }
  }
}
```

```js
---
filename: client.js
---
export default {
  async fetch(request, env) {
    using foo = await env.MY_SERVICE.foo();
    let baz = await foo.bar.baz();
    return new Response(baz);
  }
}
```

If the initial RPC ends up throwing an exception, then any pipelined calls will also fail with the same exception

> ## Proxying and forwarding stubs

Reviewer 2: Not sure this is lifecycle-related either, but doesn't seem worth its own page. I'd either move it down so that `dup()` comes first, or maybe this should go in "compatible-types" (it's talking about what happens when you send a stub across RPC...)
